std/array.d

```diff
@@ -3927,3 +3927,218 @@ unittest
     assert(appS.data == "hellow");
     assert(appA.data == "hellow");
 }
+
+/++
+Constructs a static array from `a`.
+The type of elements can be specified implicitly (`[1,2].staticArray` of type int[2])
+or explicitly (`[1,2].staticArray!float` of type float[2]).
+When `a` is a range (not known at compile time), the number of elements has to be given as template argument
+(e.g. `myrange.staticArray!2`).
+Size and type can be combined (eg: `2.iota.staticArray!(byte[2])`).
+When the range `a` is known at compile time, it can also be specified as a
+template argument to avoid having to specify the number of elements
+(eg: `staticArray!(2.iota)` or `staticArray!(double, 2.iota)`).
+
+Note: `staticArray` returns by value, so expressions involving large arrays may be inefficient.
+
+Params: a = The input elements
+
+Returns: A static array constructed from `a`.
++/
+
+pragma(inline, true) T[n] staticArray(T, size_t n)(auto ref T[n] a) nothrow @safe pure @nogc
+{
+    return a;
+}
+
+/// ditto
+U[n] staticArray(U, T, size_t n)(auto ref T[n] a) nothrow
```

Comment:
I'd accept the argument "let's leave it nothrow and worry about it later) if (a) this PR were urgent, and/or (b) the right code would be difficult to implement. Neither is applicable. The right code is (stylized):

```d
U[n] staticArray(U, T, size_t n)(auto ref T[n] a)
{
    union Temp { U[n] payload; }
    Temp result = void;
    size_t i = 0;
    static if (hasElaborateDestructor!U) scope (exit)
    {
        foreach (j; 0 .. i)
            destroy(result.payload[j]);
    }
    for (; i != n; ++i)
        emplaceRef!U(result.payload[i], cast(U) a[i]);
    return result.payload;
}
```

That's awkward and disables named return value optimization, so here's what I think would be a better approach:

```d
U[n] staticArray(U, T, size_t n)(auto ref T[n] a)
{
    U[n] result = void;
    size_t i = 0;
    static if (hasElaborateDestructor!U) scope (failure)
    {
        foreach (j; i .. n)
            emplace(result[j]); // initialize with T.init to make sure is destroyable
    }
    for (; i != n; ++i)
        emplaceRef!U(result[i], cast(U) a[i]);
    return result;
}
```

This is cute and hopefully not too much so - it comes from the opposite direction: in case of an exception, it initializes (in a nothrow manner) the uninitialized part of the result and then leaves the rest to the compiler. This allows NRVO, is efficient on the nothrow path, and is simple.